core/server/master/src/main/java/alluxio/master/file/meta/AsyncUfsAbsentPathCache.java

```diff
@@ -0,0 +1,234 @@
+/**
+ * This is a {@link UfsAbsentPathCache} which implements asynchronous addition and removal to the
+ * cache, since the processing of the path may be slow.
+ */
+@ThreadSafe
+public final class AsyncUfsAbsentPathCache implements UfsAbsentPathCache {
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncUfsAbsentPathCache.class);
+  /** Number of threads for the async pool. */
+  private static final int NUM_THREADS = 50;
+  /** Number of seconds to keep threads alive. */
+  private static final int THREAD_KEEP_ALIVE_SECONDS = 60;
+  /** Number of paths to cache. */
+  private static final int MAX_PATHS =
+      Configuration.getInt(PropertyKey.MASTER_UFS_PATH_CACHE_CAPACITY);
+
+  /** The mount table. */
+  private final MountTable mMountTable;
+  /** Paths currently being processed. This is used to prevent duplicate processing. */
+  private final ConcurrentHashMapV8<String, ReadWriteLock> mCurrentPaths;
+  /** Cache of paths which are absent in the ufs. */
+  private final Cache<String, Long> mCache;
+  /** A thread pool for the async tasks. */
+  private final ThreadPoolExecutor mPool;
+
+  /**
+   * Creates a new instance of {@link AsyncUfsAbsentPathCache}.
+   *
+   * @param mountTable the mount table
+   */
+  public AsyncUfsAbsentPathCache(MountTable mountTable) {
+    mMountTable = mountTable;
+    mCurrentPaths = new ConcurrentHashMapV8<>(8, 0.95f, 8);
+    mCache = CacheBuilder.newBuilder().maximumSize(MAX_PATHS).build();
+
+    mPool = new ThreadPoolExecutor(NUM_THREADS, NUM_THREADS, THREAD_KEEP_ALIVE_SECONDS,
+        TimeUnit.SECONDS, new LinkedBlockingQueue<Runnable>(),
+        ThreadFactoryUtils.build("UFS-Absent-Path-Cache-%d", true));
+  }
+
+  @Override
+  public void process(AlluxioURI path) {
+    mPool.submit(new ProcessPathTask(path));
+  }
+
+  @Override
+  public boolean isAbsent(AlluxioURI path) {
+    MountInfo mountInfo = getMountInfo(path);
+    if (mountInfo == null) {
+      return false;
+    }
+    AlluxioURI mountBaseUri = mountInfo.getAlluxioUri();
+
+    while (path != null && !path.equals(mountBaseUri)) {
+      Long cached = mCache.getIfPresent(path.getPath());
+      if (cached != null && cached == mountInfo.getMountId()) {
+        return true;
+      }
+      path = path.getParent();
+    }
+    // Reached the root, without finding anything in the cache.
+    return false;
+  }
+
+  /**
+   * Checks the existence of the corresponding ufs path for the given Alluxio path.
+   *
+   * @param alluxioUri the Alluxio path to check
+   * @param mountInfo the associated {@link MountInfo} for the Alluxio path
+   * @return if true, further traversal of the descendant paths should continue
+   */
+  private boolean checkPath(AlluxioURI alluxioUri, MountInfo mountInfo) {
+    ReadWriteLock rwLock = new ReentrantReadWriteLock();
+    Lock writeLock = rwLock.writeLock();
+    Lock readLock = null;
+    try {
+      // Write lock this path, to only enable a single task per path
+      writeLock.lock();
```

Comment:
at this point the lock is a local variable, so it will not achieve mutual exclusion